### ISSUE TITLE
Duplicate symbol on projects with TwilioSDK

### ISSUE THREAD

**Issue by betzerra** [mapbox/DEPRECATED-mapbox-ios-sdk#342]
I tried to install using the Binary approach and I got this error when I try to run using the simulator:

```
duplicate symbol _pj_init in:
    /Users/betzerra/Documents/Coding/ios-app/MyApp/MyApp/Libs/MapBox/MapBox.framework/MapBox(pj_init.o)
    /Users/betzerra/Documents/Coding/ios-app/MyApp/Pods/TwilioSDK/Libraries/libTwilioClient.a(os_core_unix.o)
ld: 1 duplicate symbol for architecture i386
```

If I use a device I won't get this compile error but the app will crash with a BAD_ACCESS when trying to exec pj_init

Any ideas?

**Comment by incanus** [comment]
Totally weird. `pj` in ours refers to Proj4, the map projection library, and I don't think Twilio has that, too. Must be a duplicate name for unrelated code. 

Hrm. 

At this point I would recommend also sending a bug in to Twilio. I'd be curious to see what their symbol is referring to and if this looks legit on their end or if it sounds like some sort of problem local to you. 

Sorry for the trouble :frowning: It's not usually this hard!

**Comment by betzerra** [comment]
@incanus thanks a lot again. I've already send them an email, I'll keep you posted :-)

**Comment by springmeyer** [comment]
Looks like they are using `pjsip` and the clash is coming from http://svn.pjsip.org/repos/pjproject/trunk/pjlib/src/pj/os_core_unix.c.

One potential route would be to upgrade to proj4 4.8 and use the `pj_init_plus` function instead: http://trac.osgeo.org/proj/wiki/ProjAPI. Though the `pj_init` symbol may still be pulled in...

**Comment by incanus** [comment]
Thanks, nice work @springmeyer. I'm hesitant to touch Proj4 for reasons we've discussed before (see also https://github.com/Alpstein/route-me/issues/85) but this is kind of a stopper. I'll see what Twilio has to say and if there are any suggestions, but it sounds like one side is going to have to bite the bullet and ours might be the easier path...

**Comment by betzerra** [comment]
Ok, so Twilio responded to my submitted ticket:

```
Hi Ezequiel,

Thanks for your patience while our engineering team reviewed your ticket. 
Our engineering team has noted the conflict, but unfortunately we aren't likely to 
be able to make SDK changes in the near term to resolve the issue. 
The quickest route to work around the issue may be to change the method in 
the other library. Apologies that we don't have a quick fix here; these kinds 
of conflicts can be a pain in iOS development.

Your MapBox project looks really neat. If you'd like to share any details about 
your integration we'd love to hear, or if theres any other way we can 
help you out please let us know.

Sincerely,
Brian
www.twilio.com
```

I guess that changing the method's name could be a pain in the ass. Right?

**Comment by incanus** [comment]
> I guess that changing the method's name could be a pain in the ass. Right?

Yeah :smile: But I think we can make this work quickly for your case. I can try renaming just that method in a branch of the SDK based on `develop` (or `release`, if you prefer) that you can run from source. This will also require moving Proj4 back into the SDK as a subproject instead of a prebuilt static library, which has the effect of making the whole SDK take longer to build as Proj4 is bigger than the rest of the SDK combined! But I think we can make this work for you in the short term. 

Will try to get to this today.

**Comment by betzerra** [comment]
Thank you so much!

**Comment by incanus** [comment]
Ok @betzerra, give the `proj-symbol-fix` branch a try. It's based off of `release` and includes a simple rename of the `pj_init` function (to `pj_init_alt`) which solves the build in my setup. I added the Twilio SDK, got the same error as you, made the fix, and got my iOS project to build just fine. 

This should be pretty similarly easy to patch against `develop` if you'd rather that branch instead. Just let me know.

**Comment by betzerra** [comment]
Release is perfect. I'll try this fix as soon as I can.
Thanks a lot!

On Tue, Nov 26, 2013 at 2:18 PM, Justin R. Miller
notifications@github.comwrote:

> Ok @betzerra https://github.com/betzerra, give the proj-symbol-fixbranch a try. It's based off of
> release and includes a simple rename of the pj_init function (to
> pj_init_alt) which solves the build in my setup. I added the Twilio SDK,
> got the same error as you, made the fix, and got my iOS project to build
> just fine.
> 
> This should be pretty similarly easy to patch against develop if you'd
> rather that branch instead. Just let me know.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/mapbox/mapbox-ios-sdk/issues/342#issuecomment-29322816
> .

## 

## 

Ezequiel Becerra
http://www.betzerra.com.ar
tw: @betzerra

**Comment by incanus** [comment]
Any update here? I'm inclined to leave this in a custom branch since: 
1. This just affects Twilio integration in the same project, which is relatively rare compared to the universe of third-party libraries. 
2. Our tile renderer is getting some love in private development right now and it or a future replacement will likely obsolete Proj4 use in this project. 

Closing for now; can re-open if the case makes sense.

**Comment by betzerra** [comment]
Justin, thank you so much again for your help.
Totally agree with you, specially because we're thinking about not using
Twilio any more :-)

I'll let you know if we need something else from you guys. But for now
we're good.

On Thu, Dec 26, 2013 at 8:32 PM, Justin R. Miller
notifications@github.comwrote:

> Closed #342 https://github.com/mapbox/mapbox-ios-sdk/issues/342.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/mapbox/mapbox-ios-sdk/issues/342
> .

## 

## 

Ezequiel Becerra
http://www.betzerra.com.ar
tw: @betzerra
